### PR TITLE
[Fixes #8406] Added `format` option to ActiveField's textInput and textarea

### LINKED ISSUE #8406: Format decimal values in an activeForm field
As far as I can see there are basicly two ways to format a decimal value for activeForm fields.

First one is to create getter und setter functions to prepare and return a formatted value. Here you can use the php NumberFormatter class which will use the language depended decimal sign and group separator. This way you need to write these functions for each decimal field (or one magic method) . 

The second way is using the maskedInput widget. Here you need to define the groupSeparator and radixPoint manually. I propose to set the default values for those fields within the widget based on the language.

A third option could be implemented similar to the widget function of the activeField class. You might allow using the formatter class or any other function which formats the output string.

### PR DESCRIPTION
**Pull request by freezysko**
Fixes #8406

### CODE REVIEW

**Comment by ConsultJL** [comment]
@freezy-sk just got this issue from Code Triage but it appears you fixed it, however you got have to merge up.

**Comment by freezy-sk** [comment]
Rebased to latest master and updated `@since` tags

File: framework/widgets/ActiveField.php [hunk 1]
```diff
@@ -679,6 +698,18 @@ protected function adjustLabelFor($options)
     }
 
     /**
+     * Set formatter from related form to options when format is specified
+     * @param array $options the input options
+     * @since 2.0.7
+     */
+    protected function adjustFormatter(&$options)
+    {
+        if (!empty($options['format'])) {
```
Comment (SilverFire) [inline framework/widgets/ActiveField.php]: `!isset`

File: framework/widgets/ActiveForm.php [hunk 2]
```diff
@@ -162,6 +164,13 @@ class ActiveForm extends Widget
      * @internal
      */
     public $attributes = [];
+    /**
+     * @var array|Formatter the formatter used to format ActiveField values.
```
Comment (SilverFire) [inline framework/widgets/ActiveForm.php]: [[ActiverField]]

File: framework/widgets/ActiveForm.php [hunk 3]
```diff
@@ -175,6 +184,15 @@ class ActiveForm extends Widget
      */
     public function init()
     {
+        if ($this->formatter == null) {
```
Comment (SilverFire) [inline framework/widgets/ActiveForm.php]: ===

File: framework/widgets/ActiveField.php [hunk 4]
```diff
@@ -434,7 +446,13 @@ public function fileInput($options = [])
      * @param array $options the tag options in terms of name-value pairs. These will be rendered as
      * the attributes of the resulting tag. The values will be HTML-encoded using [[Html::encode()]].
      *
-     * If you set a custom `id` for the textarea element, you may need to adjust the [[$selectors]] accordingly.
+     * The following special options are recognized:
+     *
+     * - format: string|array in which format should the value be displayed as (e.g. `"raw"`, `"text"`, `"html"`,
```
Comment (SilverFire) [inline framework/widgets/ActiveField.php]: <code>`format`</code>

File: framework/widgets/ActiveField.php [hunk 5]
```diff
@@ -347,6 +354,10 @@ public function input($type, $options = [])
      *   by a string validator, the `maxlength` option will take the value of [[\yii\validators\StringValidator::max]].
      *   This is available since version 2.0.3.
      *
+     * - format: string|array in which format should the value be displayed as (e.g. `"raw"`, `"text"`, `"html"`,
```
Comment (SilverFire) [inline framework/widgets/ActiveField.php]: <code>\`format\`</code>

File: framework/widgets/ActiveField.php [hunk 6]
```diff
@@ -321,7 +321,13 @@ public function hint($content, $options = [])
      * @param array $options the tag options in terms of name-value pairs. These will be rendered as
      * the attributes of the resulting tag. The values will be HTML-encoded using [[Html::encode()]].
      *
-     * If you set a custom `id` for the input element, you may need to adjust the [[$selectors]] accordingly.
+     * The following special options are recognized:
+     *
+     * - format: string|array in which format should the value be displayed as (e.g. `"raw"`, `"text"`, `"html"`,
```
Comment (SilverFire) [inline framework/widgets/ActiveField.php]: <code>\`format\`</code>

File: framework/helpers/BaseHtml.php [hunk 7]
```diff
@@ -2051,6 +2061,17 @@ public static function getAttributeValue($model, $attribute)
             return is_array($value) ? json_encode($value) : $value;
         }
 
+        // Format value
+        if (!empty($options['format']) && !empty($options['formatter'])) {
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: empty -> isset

File: framework/helpers/BaseHtml.php [hunk 8]
```diff
@@ -2017,10 +2026,11 @@ public static function getAttributeName($attribute)
      *
      * @param Model $model the model object
      * @param string $attribute the attribute name or expression
+     * @param array $options
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: We need a description of that options

File: framework/helpers/BaseHtml.php [hunk 9]
```diff
@@ -1204,8 +1205,12 @@ public static function error($model, $attribute, $options = [])
      */
     public static function activeInput($type, $model, $attribute, $options = [])
     {
+        $attributeValueOptions = [
+            'format' => ArrayHelper::remove($options, 'format'),
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: These special options must be described in PHPDoc

File: framework/helpers/BaseHtml.php [hunk 10]
```diff
@@ -2017,10 +2026,11 @@ public static function getAttributeName($attribute)
      *
      * @param Model $model the model object
      * @param string $attribute the attribute name or expression
+     * @param array $options
      * @return string|array the corresponding attribute value
      * @throws InvalidParamException if the attribute name contains non-word characters.
      */
-    public static function getAttributeValue($model, $attribute)
+    public static function getAttributeValue($model, $attribute, $options = [])
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: Method signature change must be mentioned in UPGRADE.md

**Comment by SilverFire** [comment]
@freezy-sk thank you for your contribution, but please, don't rush to resolve my comments because this PR is under discussion from now.

PR solves issue #8406 reported by @twinni. Personally I faced this problem many times in my own projects and always solved it somehow like this:

```php
$form->field($model, 'amount')->textInput(['value' => Yii::$app->formatter->asDecimal($model->value)]);
// or
$form->field($model, 'amount', ['inputOptions' => ['value' => Yii::$app->formatter->asDecimal($model->value)]]);
```

With this PR it can be done with the following code:

```php
$form->field($model, 'amount')->textInput(['format' => 'decimal']);
// or
$form->field($model, 'amount', ['inputOptions => ['format' => 'decimal']]);
```

It is definitely shorter, but do we need this kind of shortcut that gives about ~100 LOC to core, more complex methods' documentation, BC breaking and implicit?

@yiisoft/core-developers thoughts?

File: framework/helpers/BaseHtml.php (hunk 7 above)
Comment (freezy-sk) [inline framework/helpers/BaseHtml.php]: @SilverFire why should i change this one?

**Comment by klimov-paul** [comment]
As I already mentioned at https://github.com/yiisoft/yii2/issues/8406#issuecomment-199337424 , the whole idea is wrong.
Overcomplicating `yii\helpers\Html` is not a solution for the problem described at #8406.

**Comment by SilverFire** [comment]
@freezy-sk Thank you for your effort. We can not accept all ideas because of design principles, therefore I'm closing this PR. 

If anybody have thought why should we merge this PR, or you have an idea how to make it better and avoid possible problems - feel free to share, the conversation is not locked.

Sicerly

**Pull request closed without merging**